overlord/fdestate/handlers.go

```diff
@@ -19,47 +19,177 @@
-	//     reduces the possible states we could end up in.
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	// XXX: unlock state and let conflict detection handle the rest?
+
+	// we only care about current key slots because this might be
+	// a re-run due a force reboot or abrupt shutdown, so we want
+	// to continue deleting the remaining key slots.
+	currentKeyslots, _, err := m.GetKeyslots(keyslotRefs)
+	if err != nil {
+		return fmt.Errorf("failed to find key slots: %v", err)
+	}
+
+	for _, keyslot := range currentKeyslots {
+		if err := secbootDeleteContainerKey(keyslot.devPath, keyslot.Name); err != nil {
+			// XXX: keep going and report errors afterwards?
+			return fmt.Errorf("failed to remove key slot %s: %v", keyslot.Ref().String(), err)
+		}
+	}
+
+	// XXX: request reboot to acconut for the case where the unlock key
+	// in the kernel keyring is one of the deleted key slots?
 	return nil
 }
 
 func (m *FDEManager) doRenameKeys(t *state.Task, tomb *tomb.Tomb) error {
-	// TODO:FDEM: implement recovery key renaming, this is currently only a
-	// mock task for testing.
-
-	// TODO:FDEM:
-	//   - this might be a re-run, make task idempotent to be reselient to
-	//     abrupt reboot/shutdown.
-	//   - distinguish between errors (undo) and pure-reboots (re-run).
-	//   - conflict detection for key slot tasks is important because it
-	//     reduces the possible states we could end up in.
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	var renames map[string]string
+	if err := t.Get("renames", &renames); err != nil {
+		return err
+	}
+
+	for _, keyslotRef := range keyslotRefs {
+		if _, ok := renames[keyslotRef.String()]; !ok {
+			return fmt.Errorf("internal error: cannot find mapping for %s", keyslotRef.String())
+		}
+	}
+
+	// XXX: unlock state and let conflict detection handle the rest?
+
+	// we only care about current key slots because this might be
+	// a re-run due a force reboot or abrupt shutdown, so we want
+	// to continue renaming the remaining key slots.
+	currentKeyslots, _, err := m.GetKeyslots(keyslotRefs)
+	if err != nil {
+		return fmt.Errorf("failed to find key slots: %v", err)
+	}
+
+	// check that all remaining renames do not already exist to
+	// prevent failing midway when doing the actual renaming below.
+	if len(currentKeyslots) != 0 {
+		var renamedKeyslotRefs []KeyslotRef
+		for _, keyslot := range currentKeyslots {
+			refKey := keyslot.Ref().String()
+			renamedRef := KeyslotRef{ContainerRole: keyslot.ContainerRole, Name: renames[refKey]}
+			renamedKeyslotRefs = append(renamedKeyslotRefs, renamedRef)
+		}
+		currentRenamedKeyslots, _, err := m.GetKeyslots(renamedKeyslotRefs)
+		if err != nil {
+			return fmt.Errorf("failed to find key slots: %v", err)
```

Comment:
same about logging instead of failing

Reply:
keyslots that don't exist are indeed ignored by design of `GetKeyslots` since we are looping through `currentKeyslots` while ignoring `missing` references.

https://github.com/canonical/snapd/blob/982e0cb88bec921e709c7b2bffaa98f049ecb0c4/overlord/fdestate/fdemgr.go#L351

Reply:
then I think the use of "find" in the error is very confusing, because basically GetKeyslots will not fail if it doesn't find things, you need some other verb in the error, also as discussed "failed to" => "cannot"